### ISSUE TITLE
Problems in MDP class from mdp.py

### ISSUE THREAD

**Issue by thundergolfer** [aimacode/aima-python#917]
### Description

Today I picked up the `mdp.py` code to assist some work with the *Value Iteration* algorithm, and I found what I think are deficiencies with the `MDP` class and the `value_iteration()` function, particularly with respect to how they handle terminal states.

* `MDP.actions(self, state)` returns `[None]` if `state` is a terminal. This is wrong, as action needs to act as a dictionary lookup key, and `None` cannot act as a key. The return value should probably be `[]`, the empty list.
* In `value_iteration()`, if the state being updated is terminal (and the above problem is fixed), then the `max()` function is called with an empty list, which throws a `ValueError`. There needs to be `default=0` added to the parameter list of `max()` to handle updating terminal states.

There is also this problem: 

* If `delta` is equal to `0`, which can happen when `gamma=1`, then the following loop breaking condition never executes: `delta < epsilon*(1 - gamma)/gamma`. Note that `gamma=1` is a valid value, so this behaviour is a bug. It can at least be fixed by adding `delta is 0 or ...` to the condition 

### Appendix: 

Code setup that discovered these problems:

```python
act_list = ['N', 'S', 'E', 'W']
terminals = [(1,1), (1,3), (3,1)]
transitions = {
    (1, 1): {
        'N': [(0.7, (1, 1)), (0.3, (1, 1))], 'S': [(0.8, (2,1)), (0.2, (1,2))],
        'E': [(0.6, (1, 2)), (0.4, (2,1))], 'W': [(0.3, (1, 1)), (0.7, (1,1))]
    },
    (1, 2): {
        'N': [(0.7, (1,2)), (0.3, (1,1))],
        'S': [(0.8, (2,2)), (0.2, (1,3))],
        'E': [(0.6, (1,3)), (0.4, (2,2))],
        'W': [(0.7, (1,1)), (0.3, (1,2))]
    },
    (1, 3): {
        'N': [(0.7, (1,3)), (0.3, (1,2))],
        'S': [(0.8, (2,3)), (0.2, (1,3))],
        'E': [(0.6, (1,3)), (0.4, (2,3))],
        'W': [(0.7, (1,2)), (0.3, (1,3))]
    },
    (2, 1): {
        'N': [(0.7, (1,1)), (0.3, (2,1))],
        'S': [(0.8, (3,1)), (0.2, (2,2))],
        'E': [(0.6, (2,2)), (0.4, (3,1))],
        'W': [(0.7, (2,1)), (0.3, (1,1))]
    },
    (2, 2): {
        'N': [(0.7, (1,2)), (0.3, (2,1))],
        'S': [(0.8, (3,2)), (0.2, (2,3))],
        'E': [(0.6, (2,3)), (0.4, (3,2))],
        'W': [(0.7, (2,1)), (0.3, (1,2))]
    },
    (2, 3): {
        'N': [(0.7, (1,3)), (0.3, (2,2))],
        'S': [(0.8, (3,3)), (0.2, (2,3))],
        'E': [(0.6, (2,3)), (0.4, (3,3))],
        'W': [(0.7, (2,2)), (0.3, (1,3))]
    },
    (3, 1): {
        'N': [(0.7, (2,1)), (0.3, (3,1))],
        'S': [(0.8, (3,1)), (0.2, (3,2))],
        'E': [(0.6, (3,2)), (0.4, (3,1))],
        'W': [(0.7, (3,1)), (0.3, (2,1))]
    },
    (3, 2): {
        'N': [(0.7, (2,2)), (0.3, (3,1))],
        'S': [(0.8, (3,2)), (0.2, (3,3))],
        'E': [(0.6, (3,3)), (0.4, (3,2))],
        'W': [(0.7, (3,1)), (0.3, (2,2))]
    },
    (3, 3): {
        'N': [(0.7, (2,3)), (0.3, (3,2))],
        'S': [(0.8, (3,3)), (0.2, (3,3))],
        'E': [(0.6, (3,3)), (0.4, (3,3))],
        'W': [(0.7, (3,2)), (0.3, (2,3))]
    }
}

rewards = {
    (1, 1): 20,
    (1, 2): -1,
    (1, 3): 5,
    (2, 1): -1,
    (2, 2): -1,
    (2, 3): -1,
    (3, 1): -20,
    (3, 2): -1,
    (3, 3): -1
}
states = list(rewards.keys())
gamma = 1
init = (1, 1)

from mdp import MDP

problem = MDP(
    init,
    act_list,
    terminals,
    transitions,
    rewards,
    states,
    gamma
)

from mdp import value_iteration

u = value_iteration(problem)

print(u)
```

**Comment by ad71** [comment]
Thanks for bringing this up. The `mdp` class was written to serve as a base class and it is mostly used after defining a custom class for complicated problems, but I agree that some of these issues make the class pretty useless. The Gamma = 1 problem exists because `value_iteration` exactly follows the pseudocode from the book. I will look into this problem and make the required changes.

**Comment by ad71** [comment]
This issue has now been fixed in #918 
The first two points were taken care of by the new `MDP2` class (I couldn't think of a better name).
The only change is that this class returns a transition probability of 0 when the action is `None`. This solves the problem without having to change the `value_iteration` function or the `MDP` class.
The third point was taken care of by replacing `<` with `<=` in the loop termination condition.

I ran your setup on this code as follows
```python
actlist = ['N', 'S', 'E', 'W']
terminals = [(1,1), (1,3), (3,1)]
transitions = {
    (1, 1): {
        'N': [(0.7, (1, 1)), (0.3, (1, 1))], 'S': [(0.8, (2,1)), (0.2, (1,2))],
        'E': [(0.6, (1, 2)), (0.4, (2,1))], 'W': [(0.3, (1, 1)), (0.7, (1,1))]
    },
    (1, 2): {
        'N': [(0.7, (1,2)), (0.3, (1,1))],
        'S': [(0.8, (2,2)), (0.2, (1,3))],
        'E': [(0.6, (1,3)), (0.4, (2,2))],
        'W': [(0.7, (1,1)), (0.3, (1,2))]
    },
    (1, 3): {
        'N': [(0.7, (1,3)), (0.3, (1,2))],
        'S': [(0.8, (2,3)), (0.2, (1,3))],
        'E': [(0.6, (1,3)), (0.4, (2,3))],
        'W': [(0.7, (1,2)), (0.3, (1,3))]
    },
    (2, 1): {
        'N': [(0.7, (1,1)), (0.3, (2,1))],
        'S': [(0.8, (3,1)), (0.2, (2,2))],
        'E': [(0.6, (2,2)), (0.4, (3,1))],
        'W': [(0.7, (2,1)), (0.3, (1,1))]
    },
    (2, 2): {
        'N': [(0.7, (1,2)), (0.3, (2,1))],
        'S': [(0.8, (3,2)), (0.2, (2,3))],
        'E': [(0.6, (2,3)), (0.4, (3,2))],
        'W': [(0.7, (2,1)), (0.3, (1,2))]
    },
    (2, 3): {
        'N': [(0.7, (1,3)), (0.3, (2,2))],
        'S': [(0.8, (3,3)), (0.2, (2,3))],
        'E': [(0.6, (2,3)), (0.4, (3,3))],
        'W': [(0.7, (2,2)), (0.3, (1,3))]
    },
    (3, 1): {
        'N': [(0.7, (2,1)), (0.3, (3,1))],
        'S': [(0.8, (3,1)), (0.2, (3,2))],
        'E': [(0.6, (3,2)), (0.4, (3,1))],
        'W': [(0.7, (3,1)), (0.3, (2,1))]
    },
    (3, 2): {
        'N': [(0.7, (2,2)), (0.3, (3,1))],
        'S': [(0.8, (3,2)), (0.2, (3,3))],
        'E': [(0.6, (3,3)), (0.4, (3,2))],
        'W': [(0.7, (3,1)), (0.3, (2,2))]
    },
    (3, 3): {
        'N': [(0.7, (2,3)), (0.3, (3,2))],
        'S': [(0.8, (3,3)), (0.2, (3,3))],
        'E': [(0.6, (3,3)), (0.4, (3,3))],
        'W': [(0.7, (3,2)), (0.3, (2,3))]
    }
}

rewards = {
    (1, 1): 20,
    (1, 2): -1,
    (1, 3): 5,
    (2, 1): -1,
    (2, 2): -1,
    (2, 3): -1,
    (3, 1): -20,
    (3, 2): -1,
    (3, 3): -1
}

gamma = 1
init = (1, 1)

from mdp import MDP2
from mdp import value_iteration

problem = MDP2(
    init,
    actlist,
    terminals,
    transitions,
    rewards,
    gamma
)

u = value_iteration(problem)
print(u)
```
and got the following result
```python
>>> {(3, 2): 8.990476190476183, (1, 3): 5.0, (2, 1): 18.57142857142857, (2, 3): 12.799999999999997, (1, 2): 18.57142857142857, (3, 3): 10.657142857142851, (3, 1): -20.0, (2, 2): 17.57142857142857, (1, 1): 20.0}
```
Hope this solves the problem.

**Comment by thundergolfer** [comment]
Thanks for your work on this. Happy for it to be closed
